### PR TITLE
Setup Java 21 container

### PR DESCRIPTION
**Pull request by Abacn**
Part of #28120

After this PR, beam will start publishing java21 dev sdk container. A wordcount compiled with Java21 locally and run on Dataflow v2 with this container has passed.

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by Abacn** [comment]
Example wordcount job (maven project) on Dataflow: https://console.cloud.google.com/dataflow/jobs/us-central1/2023-10-05_17_20_31-16427189156223040319?project=google.com%3Aclouddfe

image

However not able to setup Java21 test within beam repo yet as gradle not yet support Java 21. Currently Java invocation fails with
```
Unrecognized option: --add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```
but if not providing these args, it fails with
```
warning: The Checker Framework is tested with JDK 8, 11, 17, and 18. You are using version 21.
error: warnings found and -Werror specified
/Users/yathu/beam/sdks/java/testing/jpms-tests/src/main/java/org/apache/beam/sdk/jpmstests/WordCount.java:38: error: [type.checking.not.run] NullnessChecker did not run because of a previous error issued by javac
public class WordCount {

```

File: .github/workflows/beam_LoadTests_Java_GBK_Dataflow_V2_Batch_Java11.yml [hunk 1]
```diff
@@ -95,7 +95,7 @@ jobs:
             -PloadTest.mainClass=org.apache.beam.sdk.loadtests.GroupByKeyLoadTest \
             -Prunner=:runners:google-cloud-dataflow-java \
             -Prunner.version=V2 \
-            -PcompileAndRunTestsWithJava11 \
+            -PtestJavaVersion=11 \
```
Comment (Abacn) [inline .github/workflows/beam_LoadTests_Java_GBK_Dataflow_V2_Batch_Java11.yml:98]: this flag change is to address https://github.com/apache/beam/pull/28726#discussion_r1341443087 (later planned as part of Java21 support: https://github.com/apache/beam/issues/28120#issuecomment-1741033265)

File: sdks/java/container/java21/option-jamm.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (Abacn) [inline sdks/java/container/java21/option-jamm.json:1]: although same options are added as java17, it seems jamm still not working on Java21:

```
JVM prevents jamm from accessing subgraph - cache sizes may be underestimated
```

support left as next task

File: sdks/java/container/Dockerfile [hunk 3]
```diff
@@ -15,8 +15,9 @@
 #  See the License for the specific language governing permissions and
 # limitations under the License.
 ###############################################################################
+ARG base_image
 ARG java_version
-FROM eclipse-temurin:${java_version}
+FROM ${base_image}:${java_version}
```
Comment (Abacn) [inline sdks/java/container/Dockerfile:20]: we switched from openjdk to temurin in #25402 due to that openjdk image for java8 and 11 was deprecated. For now we still use temurin for 8,11,17; openjdk for 21 temporarily until upstream https://adoptium.net/blog/2023/09/temurin21-delay/ resolved

**Review by kennknowles — APPROVED** [review]

**Comment by Abacn** [comment]
Because this PR changed test parameters (Jenkins, GHA), should be merged after 2.51.0 release finalized

**Comment by Abacn** [comment]
Note that temurin java21 has released in Oct 10th: https://adoptium.net/temurin/releases/

**Comment by Abacn** [comment]
Run Java PreCommit

**Comment by Abacn** [comment]
Run PythonDocker PreCommit

**Comment by Abacn** [comment]
Run Python_PVR_Flink PreCommit

**Comment by Abacn** [comment]
Run Java_PVR_Flink_Docker PreCommit

**Comment by Abacn** [comment]
tests either passed on GHA or Jenkins, merging

**Pull request merged**